larq_compute_engine/tflite/kernels/bconv2d.cc

```diff
@@ -125,12 +126,25 @@ void* Init(TfLiteContext* context, const char* buffer, size_t length) {
     conv_params->dilations[i] = dilation_vector[i].AsInt64();
 
   // reading padding
-  conv_params->padding_type =
-      m["padding"].ToString() == "VALID" ||
-              m["padding"].ToString() ==
-                  "valid"  // TODO: not sure if this check is needed
-          ? TfLitePadding::kTfLitePaddingValid
-          : TfLitePadding::kTfLitePaddingSame;
+  if (m["padding"].ToString() == "VALID" ||
+      m["padding"].ToString() == "valid") {
+    conv_params->padding_type = kTfLitePaddingValid;
+  } else if (m["padding"].ToString() == "SAME" ||
+             m["padding"].ToString() == "same") {
+    conv_params->padding_type = kTfLitePaddingSame;
+  } else {
+    context->ReportError(context, "Invalid padding attribute.");
+    return conv_params;
+  }
+
+  conv_params->pad_value =
+      m["pad_values"].IsNull() ? 0 : m["pad_values"].AsInt64();
+
+  if (conv_params->pad_value != 0 && conv_params->pad_value != 1) {
```

Comment:
1. the pad value can be any number (not only 0 and 1) and during the bitpacking, we automatically pack it to the right bit (0 is the problematic pad value which we need to check explicitly). 
2. currently we don't have negative numbers padding support and thats why we have this check?

Reply:
`pad_value` is the value that is written in the pixels around the image by the im2col algorithm.
(i.e. its the exact same value as in the larq layers   larq/larq#438 )

The im2col algorithm is done after bitpacking, so it can only write binary 0 (i.e. float +1, `pad_value=1`) or binary 1 (i.e. float -1, `pad_value=-1`).

When `pad_value=0` then we have zero-padding, and we write `+1` around the image and then have the zero-padding-correction-functor to fix it.  When `pad_value=1` then we have one-padding, so we *also* write `+1` around the image and then we dont have a correction.

We could also add support for `pad_value=-1` but the MLIR converter will always give us models with `+1` so i think this is cleaner.

Reply:
what about the cases we do the im2col before bitpacking?

Reply:
In that case, im2col happens in "float space" and we could choose which value we want to pad with (the `zero_byte = 0x00`). But then the bitpacking is applied and it will turn it into binary 0 or 1. So in that case we have effectively padded with -1 or +1 as well.